Update Lists.scala

nishanpatel opened the pull request:
As map, flatMap and filter are very useful methods for list so thought to have in exercise.

nishanpatel commented:
Hi, @juanpedromoreno & @rafaparadela, would you mind to review PR created and provide your feedback? Please let me know or decline if it is not relevant to project.

nishanpatel commented:
Hi Frank,

I thing it is good idea to split exercise based on function level. Let me see how I can split into different ones and share it with you.

nishanpatel commented:
Hi Frank,

As discussed I have updated PR and split the exercise based on functional level(reverse, filter, map, flatten and flatMap). Could you please verify and let me know in case of any issue?

Thanks,
Nishan

nishanpatel commented:
> Sure @nishanpatel I'll take a look asap. I was a bit busy the last days so please accept my apologies for not being very responsive!

Hi @FRosner , don't be apologies, take your time and I have just message you so that you will get notification and review whenever you get some time review the same. 

Thanks,
Nishan

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
-  def usefulMethodsLists(res0: Int, res1: List[Int], res2: List[Int], res3: List[Int]) {
+  /** Lists have many useful utility methods.
+    *
+    * `reverse` return a reversed copy of the list:
```
> `reverse` returns a reversed copy of the list:

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
-    a.map { v ⇒
-      v * 2
-    } should equal(res2)
+  /** `filter` tests each element of the List by applying a function you supply:
```
How about "`filter` discards each element of the list that does not satisfy the provided filter condition:"?

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
     val a = List(1, 3, 5, 7, 9)
 
     // get the length of the list
     a.length should equal(res0)
```
I think we shouldn't include `length` into the same exercise as `reverse`, or? What do you think?

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
+    } should equal(res1)
+
+    // Functions over lists can use () instead of {}
+    a.filter(_ % 2 != 0) should equal(res2)
```
Just to remind me: Was the difference between `()` and `{}` introduced somewhere else already? If possible I'd not mix the two concepts here and only focus on new things related to lists. What do you think?

nishanpatel replied:
Yes we had duplicate reference  for  _ as shorthand and () instead of {} in map method. I have removed duplicate code reference from map method.

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
+
+    a.flatten should equal(res0)
+
+    b.map { v =>
```
I think we don't necessarily have to duplicate the map.flatten = flatMap example here. What do you think?

nishanpatel replied:
I think it will be better to have both the example here so that person can understand easily and co-related functions easily.

FRosner commented on src/main/scala/stdlib/Lists.scala:
```diff
@@ -90,50 +90,93 @@ object Lists extends FlatSpec with Matchers with org.scalaexercises.definitions.
+  def flatMapLists(res0: List[Int], res1: Boolean) {
+    val a = List(1,2)
+
+    // The flatMap operator is similar to map, but it takes a function returning a list of elements
```
Are these inline comments shown in the UI?

nishanpatel replied:
Not sure as I am having issues with setup project local in my machine

nishanpatel commented:
Hi @FRosner thanks for providing the feedback and your time. I tried to setup the application in my local environment and followed the steps mentioned in > scala-exercises, however, I was getting 

> [error] (coreJVM/compile:createHeaders) java.nio.charset.UnmappableCharacterException: Input length = 1

exception and failed to start server. So unforgettably was not able to run in local. 

I am using following configuration. Windows 10, Scala version 2.12.6, java version 1.8.0_161 and SBT version 1.2.1. Could you please let me know am I missing something while setup in my local?

Thank you and appreciate your help.

FRosner commented:
Can you try using a lower version of SBT (1.0.x)? I remember people in the chat having the same issue. What did Google say?
-- 
Sent from my phone. Please excuse my brevity.

nishanpatel commented:
Thanks Frank. I tried with lower version of SBT but still having the same issue. Google said this issue with Play/SBT fixed already but I am facing the same issue still. [https://github.com/playframework/playframework/pull/8495](Windows Issue)

FRosner commented:
Could you search the gitter history I think people discussed it already but I'm not sure if there was a solution. Worst case try SBT 0.13?
-- 
Sent from my phone. Please excuse my brevity.On 9/18/18, 12:36 Nishan Patel <notifications@github.com> wrote:
Thanks Frank. I tried with lower version of SBT but still having the same issue. Google said this issue with Play/SBT fixed already but I am facing the same issue still. [https://github.com/playframework/playframework/pull/8495](Windows Issue)

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/scala-exercises/exercises-stdlib","title":"scala-exercises/exercises-stdlib","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/scala-exercises/exercises-stdlib"}},"updates":{"snippets":[{"icon":"PERSON","message":"@nishanpatel in #108: Thanks Frank. I tried with lower version of SBT but still having the same issue. Google said this issue with Play/SBT fixed already but I am facing the same issue still. [https://github.com/playframework/playframework/pull/8495](Windows Issue)"}],"action":{"name":"View Pull Request","url":"https://github.com/scala-exercises/exercises-stdlib/pull/108#issuecomment-422343050"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/scala-exercises/exercises-stdlib/pull/108#issuecomment-422343050",
"url": "https://github.com/scala-exercises/exercises-stdlib/pull/108#issuecomment-422343050",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [scala-exercises/exercises-stdlib] Update Lists.scala (#108)",
"sections": [
{
"text": "",
"activityTitle": "**Nishan Patel**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@nishanpatel",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"scala-exercises/exercises-stdlib\",\n\"issueId\": 108,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close pull request",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"PullRequestClose\",\n\"repositoryFullName\": \"scala-exercises/exercises-stdlib\",\n\"pullRequestId\": 108\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/scala-exercises/exercises-stdlib/pull/108#issuecomment-422343050"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 369919990\n}"
}
],
"themeColor": "26292E"
}
]

nishanpatel commented:
With SBT 0.13 also having the same issue.

FRosner commented:
I'll take a look on the weekend.
-- 
Sent from my phone. Please excuse my brevity.

nishanpatel commented:
> I'll take a look on the weekend. -- Sent from my phone. Please excuse my brevity.

Thanks Frank.

FRosner commented:
@nishanpatel can you try one of the workarounds described in https://github.com/playframework/playframework/issues/8449 and see if this solves the problem? Thanks!

nishanpatel commented:
> @nishanpatel can you try one of the workarounds described in [playframework/playframework#8449](https://github.com/playframework/playframework/issues/8449) and see if this solves the problem? Thanks!

Thanks @FRosner for the link. It was my bad I have came across this link and now issue has been resolved it seems. But now I am getting below issue.

> Exception: 
Sass compiler: 'sass.bat' is not recognized as an internal or external command,
File: C:\Users\npatel2\IdeaProjects\scala-exercises\server\app\assets\stylesheets\main.scss
Line: 0 Col: 0

I already tried solutions provided on git. Installed Sass using gem install sass but no luck.
thanks again for your time, appreciate your help.

nishanpatel commented:
Hi @FRosner thanks for help. Now I am able to run project in my local but still I won't see my changes in application. I am facing issues to build exercises-stdlib project in local and getting below issue. 

>  10 warnings found
java.util.regex.PatternSyntaxException: Illegal/unsupported escape sequence near index 3
C:\Users\npatel2\IdeaProjects\exercises-stdlib
   ^
        at java.util.regex.Pattern.error(Pattern.java:1957)
        at java.util.regex.Pattern.escape(Pattern.java:2473)
        at java.util.regex.Pattern.atom(Pattern.java:2200)
        at java.util.regex.Pattern.sequence(Pattern.java:2132)
        at java.util.regex.Pattern.expr(Pattern.java:1998)
        at java.util.regex.Pattern.compile(Pattern.java:1698)
        at java.util.regex.Pattern.<init>(Pattern.java:1351)
        at java.util.regex.Pattern.compile(Pattern.java:1028)
        at java.lang.String.split(String.java:2380)
        at java.lang.String.split(String.java:2422)
        at org.scalaexercises.compiler.SourceTextExtraction.relativePath(SourceTextExtraction.scala:66)
        at org.scalaexercises.compiler.SourceTextExtraction$$anonfun$1.apply(SourceTextExtraction.scala:74)
        at org.scalaexercises.compiler.SourceTextExtraction$$anonfun$1.apply(SourceTextExtraction.scala:73)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at org.scalaexercises.compiler.SourceTextExtraction.extractAll(SourceTextExtraction.scala:73)
        at org.scalaexercises.compiler.Compiler.compile(Compiler.scala:81)
        at org.scalaexercises.compiler.CompilerJava.compile(Compiler.scala:51)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$anonfun$$invokeCompiler$1$1$$anonfun$apply$9.apply(ExerciseCompilerPlugin.scala:219)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$anonfun$$invokeCompiler$1$1$$anonfun$apply$9.apply(ExerciseCompilerPlugin.scala:228)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$.org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$captureStdStreams(ExerciseCompilerPlugin.scala:281)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$anonfun$$invokeCompiler$1$1.apply(ExerciseCompilerPlugin.scala:217)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$anonfun$$invokeCompiler$1$1.apply(ExerciseCompilerPlugin.scala:209)
        at cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:267)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1.org$scalaexercises$plugin$sbtexercise$ExerciseCompilerPlugin$$anonfun$$invokeCompiler$1(ExerciseCompilerPlugin.scala:209)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14$$anonfun$apply$16$$anonfun$apply$17.apply(ExerciseCompilerPlugin.scala:245)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14$$anonfun$apply$16$$anonfun$apply$17.apply(ExerciseCompilerPlugin.scala:245)
        at cats.Traverse$$anonfun$traverseU$1.apply(Traverse.scala:51)
        at cats.instances.ListInstances$$anon$1$$anonfun$traverse$2.apply(list.scala:70)
        at cats.instances.ListInstances$$anon$1$$anonfun$traverse$2.apply(list.scala:69)
        at cats.instances.ListInstances$$anon$1.cats$instances$ListInstances$class$$anon$$loop$2(list.scala:58)
        at cats.instances.ListInstances$$anon$1$$anonfun$foldRight$1.apply(list.scala:60)
        at cats.instances.ListInstances$$anon$1$$anonfun$foldRight$1.apply(list.scala:60)
        at cats.Eval$Call$.cats$Eval$Call$$loop(Eval.scala:236)
        at cats.Eval$Call.value(Eval.scala:229)
        at cats.instances.ListInstances$$anon$1.traverse(list.scala:71)
        at cats.instances.ListInstances$$anon$1.traverse(list.scala:12)
        at cats.Applicative$class.traverse(Applicative.scala:35)
        at cats.instances.EitherInstances$$anon$1.traverse(either.scala:32)
        at cats.Traverse$class.traverseU(Traverse.scala:51)
        at cats.instances.ListInstances$$anon$1.traverseU(list.scala:12)
        at cats.Traverse$Ops$class.traverseU(Traverse.scala:16)
        at cats.Traverse$ToTraverseOps$$anon$4.traverseU(Traverse.scala:16)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14$$anonfun$apply$16.apply(ExerciseCompilerPlugin.scala:245)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14$$anonfun$apply$16.apply(ExerciseCompilerPlugin.scala:244)
        at cats.syntax.EitherOps$.flatMap$extension(either.scala:129)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14.apply(ExerciseCompilerPlugin.scala:244)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9$$anonfun$apply$14.apply(ExerciseCompilerPlugin.scala:241)
        at cats.syntax.EitherOps$.flatMap$extension(either.scala:129)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9.apply(ExerciseCompilerPlugin.scala:241)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1$$anonfun$9.apply(ExerciseCompilerPlugin.scala:238)
        at cats.syntax.EitherOps$.flatMap$extension(either.scala:129)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1.apply(ExerciseCompilerPlugin.scala:238)
        at org.scalaexercises.plugin.sbtexercise.ExerciseCompilerPlugin$$anonfun$generateExercisesTask$1.apply(ExerciseCompilerPlugin.scala:162)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
        at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:237)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[error] (compile-generated-exercises:generateExercises) java.util.regex.PatternSyntaxException: Illegal/unsupported escape sequence near index 3
[error] C:\Users\npatel2\IdeaProjects\exercises-stdlib
[error]    ^
[error] Total time: 60 s, completed 28-Sep-2018 16:11:13

FRosner commented:
Hi @nishanpatel

I fear I'm too busy to be responsive at the moment. Do you think you can try yourself / asking someone else for help for this problem? I'm happy to finish what I started and review and merge the PR but I can't find the time to debug the Windows setup atm.

Sorry :(
